File: showcase/src/engine/translator.ts
import { isArabic, lookupAr, lookupEn } from "./lookupIndices";

export function translateToken(token: string, to?: "ar" | "en"): string {
  const srcIsAr = isArabic(token);
  const target = to || (srcIsAr ? "en" : "ar");

  if (target === "en") {
    const entry = lookupAr(token);
    return entry ? entry.en : token;
  } else {
    const entry = lookupEn(token);
    return entry ? entry.ar : token;
  }
}

export function translate(fullName: string, to?: "ar" | "en"): string {
  if (!fullName || !fullName.trim()) return fullName;
  const tokens = fullName.trim().split(/\s+/);
  return tokens.map((t) => translateToken(t, to)).join(" ");
}

File: csharp/EgyNames/Translator.cs
using System;
using System.Linq;

namespace EgyptianNames
{
    public static class Translator
    {
        public static string TranslateToken(string token, string? to = null, string? dataPath = null)
        {
            bool srcIsAr = LookupIndices.IsArabic(token);
            string target = to ?? (srcIsAr ? "en" : "ar");

            if (target.Equals("en", StringComparison.OrdinalIgnoreCase))
            {
                var entry = LookupIndices.LookupAr(token, dataPath);
                return entry != null ? entry.En : token;
            }
            else
            {
                var entry = LookupIndices.LookupEn(token, dataPath);
                return entry != null ? entry.Ar : token;
            }
        }

        public static string Translate(string fullName, string? to = null, string? dataPath = null)
        {
            if (string.IsNullOrWhiteSpace(fullName)) return fullName;
            var tokens = fullName.Trim().Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
            var translated = tokens.Select(t => TranslateToken(t, to, dataPath));
            return string.Join(" ", translated);
        }
    }
}
